### PR TITLE
sync: return an error for rename table statement in optimistic mode

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
we do not support rename table now, simply return an error in optimistic mode.
For pessimistic mode, though shardddl group was wrong, dm works well and data are still consistent 🙈.

TODO: support rename table for all mode.
### What is changed and how it works?
prefilter before handle statement in optimistic mode.

### Check List 

Tests 
 - Integration test

TODO: emphasize don't mixed sharding tables with normal tables in one task, in other words,`sharding-mode` means applying to all tables in this task, not only table-routes tables

### CODE REVIEW

File: pkg/terror/error_list.go [hunk 1]
```diff
@@ -998,6 +999,7 @@ var (
 	ErrSyncerOperatorNotExist               = New(codeSyncerOperatorNotExist, ClassSyncUnit, ScopeInternal, LevelLow, "error operator not exist, position: %s", "")
 	ErrSyncerReplaceEventNotExist           = New(codeSyncerReplaceEventNotExist, ClassSyncUnit, ScopeInternal, LevelHigh, "replace event not exist, location: %s", "")
 	ErrSyncerParseDDL                       = New(codeSyncerParseDDL, ClassSyncUnit, ScopeInternal, LevelHigh, "parse DDL: %s", "Please confirm your DDL statement is correct and needed. For TiDB compatible DDL, see https://docs.pingcap.com/tidb/stable/mysql-compatibility#ddl. You can use `handle-error` command to skip or replace the DDL or add a binlog filter rule to ignore it if the DDL is not needed.")
+	ErrSyncerUnsupportedStmt                = New(codeSyncerUnsupportedStmt, ClassSyncUnit, ScopeInternal, LevelHigh, "`%s` statement not supported in %s mode", "")
```
Comment (lance6716) [inline pkg/terror/error_list.go:1002]: we could add a workaround 🤔

**Comment by lance6716** [comment]
need to tell user how to handle this situation. rest LGTM

File: pkg/terror/error_list.go (hunk 1 above)
Comment (GMHDBJD) [reply pkg/terror/error_list.go:1002]: I can't find out a solution. Maybe we should support rename to the same group now. 😥

File: syncer/syncer.go [hunk 2]
```diff
@@ -1837,6 +1837,10 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
 			case *ast.TruncateTableStmt:
 				ec.tctx.L().Info("ignore truncate table statement in shard group", zap.String("event", "query"), zap.String("statement", sqlDDL))
 				continue
+			// TODO: Support rename a table to the same shard group
+			// e.g. upstream: tb* -> downstream: tb, rename tb1 to tb2
+			case *ast.RenameTableStmt:
```
Comment (lance6716) [inline syncer/syncer.go:1842]: will it error for the non-sharding tables?

Comment (GMHDBJD) [reply syncer/syncer.go:1842]: Yes, it does happen with the above implementation. I'm working on “rename to the same group" and will try to fix it.

Comment (GMHDBJD) [reply syncer/syncer.go]: Actually, if mode is pessimistic, there isn't non-sharding tables.

Comment (lance6716) [reply syncer/syncer.go]: OK, that's related to https://github.com/pingcap/dm/issues/1517. In future we should support specify part of sources to form a sharding group.

File: syncer/syncer.go [hunk 3]
```diff
@@ -1856,6 +1856,8 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
 			case *ast.TruncateTableStmt:
 				ec.tctx.L().Info("ignore truncate table statement in shard group", zap.String("event", "query"), zap.String("statement", sqlDDL))
 				continue
+			case *ast.RenameTableStmt:
```
Comment (lance6716) [inline syncer/syncer.go:1859]: we heavily rely on "an optimistic sharding task should sync no other tables than `table-route`", if user both sync sharding tables and non sharding tables in one task, non sharding tables will not permitted to RENAME.

How about we give an error when user both specify block-allow list and table routes, to ensure user add a minimal sharding task that will not sync non-sharding tables? we might talk to PM if you think that's reasonable. @GMHDBJD @lichunzhu 

could join us after your vacation @gozssky

Comment (lance6716) [reply syncer/syncer.go:1859]: (but https://github.com/pingcap/dm/issues/1517 is still not solved)

Comment (GMHDBJD) [reply syncer/syncer.go:1859]: Seems there is not associated between #1517 and rename table? We will support rename in both pessimistic and optimistic later. It's no conflict that we can also support #1517 later.

Comment (lance6716) [reply syncer/syncer.go:1859]: yes, what do you think "this PR also forbid user RENAME a non sharding table in a optimistic sharding task", could we raise an error when user both specify block-allow list and table routes (use table routes as allow list)?

Comment (GMHDBJD) [reply syncer/syncer.go:1859]: We have other stmts not supported for a non sharding table in shard mode(truncate statement). So it's not a problem of Rename.
> raise an error when user both specify block-allow list and table routes 

We can support it in another pr or #1517? Maybe task to PM will be better.

**Pull request merged**